runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java

```diff
@@ -0,0 +1,304 @@
+         * @param o object to compare against.
+         * @throws IllegalArgumentException if the two Ids are not comparable.
+         * @return results of comparison.
+         */
+        @Override
+        public int compareTo(CorfuRecordId o) {
+            if (this.id.getMostSignificantBits() == 0 && o.id.getMostSignificantBits() != 0) {
+                throw new IllegalArgumentException(
+                        "Incompatible CorfuRecordId comparison: ordering unavailable");
+            }
+            if (this.id.getMostSignificantBits() !=0 && o.id.getMostSignificantBits() == 0) {
+               throw new IllegalArgumentException(
+                       "Incompatible CorfuRecordId comparison: order of compared object unknown");
+            }
+            return id.compareTo(o.id);
+        }
+
+        /**
+         * It is ok to check equality of a CorfuRecordId with ordering data against one without.
+         * @param o object to compare against.
+         * @return
+         */
+        public boolean equals(CorfuRecordId o) {
+            return id.getLeastSignificantBits() == o.id.getLeastSignificantBits();
+        }
+
+        public String toString() {
+            return id.toString();
+        }
+    }
+
+    /**
+     * Appends the specified element at the end of this unbounded queue.
+     * In a distributed system, the linearizable order of insertions cannot be guaranteed
+     * unless a transaction is used.
+     * Capacity restrictions and backoffs must be implemented outside this
+     * interface. Consider validating the size of the queue against a high
+     * watermark before enqueue.
+     *
+     * @param e the element to add
+     * @throws IllegalArgumentException if some property of the specified
+     *         element prevents it from being added to this queue
+     * @return Unique ID representing this entry in the persistent queue
+     * WARNING: The ID returned by this function will NOT be the comparable with the ID
+     *          returned by the CorfuQueueRecord from entryList() method because if this
+     *          method executes within a transaction, the comparable ordering is not
+     *          known until the transaction commits.
+     *          The ID returned here is only really useful for remove() operations.
+     */
+    public CorfuRecordId enqueue(E e) {
+        final Long id = guidGenerator.nextLong();
+        corfuTable.put(id, e);
+        return new CorfuRecordId(0, id);
+    }
+
+    /**
+     * CorfuQueueRecord encapsulates each entry enqueued into CorfuQueue with its unique ID.
+     * It is a read-only type returned by the entryList() method.
+     * The ID returned here can be used for both point get()s as well as remove() operations
+     * on this Queue.
+     *
+     * @param <E>
+     */
+    public static class CorfuQueueRecord<E> implements Comparable<CorfuQueueRecord<? extends E>> {
```

Comment:
`Comparable<CorfuQueueRecord<E>>` ?

Reply:
Comparable<CorfuQueueRecord<? extends E>> was in response to @lixinchengdu 's comments above. Please let me know if you agree? thanks

Reply:
I am not sure whether there is any pitfall by doing this way. I am checking whether other having the same issue .